### ISSUE TITLE
Routes (/test/start_test) and (/test/end_test)

### ISSUE THREAD

**Issue by RishabhVerma** [navgurukul/chanakya#24]
@amarkrsinha1997 we need to finish these ASAP. There is a certain logic that is used to generate the question paper randomly for every student. You will have to look up that logic in the old code and document it here. I don't remember the logic exactly.

Like before please document the complete implementation before you begin the code.

**Comment by amarkrsinha1997** [comment]
/test/start_test/
/GET
check for the enrollment key if it is valid then will 
start the timer in enrollment key
and will randomly select question out of the database and then will send it to the student
else it will send the invalid status

/test/end_test
/POST it will accept the question and the answer and will save all the data in QuestionAttempt model  with the enrollment key

shall we create any field to record calculated marks as it doing calculation all the time will not be good or shall we just keep it like it is

**Comment by RishabhVerma** [comment]
@amarkrsinha1997 Please document the logic on how questions would be selected randomly from the DB. Also if we're storing marks with every attempt of the student computing marks in real time shouldn't be a big issue. Let's write the code without storing marks for every attempt. We can refactor it later on.

You can find the logic in the old version of the code.

**Comment by amarkrsinha1997** [comment]
should there be a config for difficulty and the categories of the student to select the question randomly according to category and it's difficulty level? is it the whole logic for the question route?

**Comment by RishabhVerma** [comment]
Yes that's the whole logic. But please make sure that you are able to document the whole logic in plain english here (by looking at the old code) before you start writing the code. That will help you gain more clarity :)

**Comment by amarkrsinha1997** [comment]
I got to logic to do this stuff which one would you prefer wanted to understand and why?

1. 
Get all the questions by querying the database at once and loop over to and add it to a dictionary containing a list example questions[topic][difficulty] and append the question in it and then randomly shuffle all the question once and then slice the question by the number of question required that is in config file.

2. Query the database as many times as the topic and difficulty are there in a config file and then shuffle the question, slice it with the number of question required in this topic and difficulty and then add it to a dictionary containing question according to difficulty and topic

which would you prefer? currently, I am implementing number 1 logic @RishabhVerma

**Comment by amarkrsinha1997** [comment]
Need a clarity on a particular part to suppose a child is giving the test and he refreshes the page while the test is going if we give him the randomized question again in the middle of the test time won't it be a problem that he has to give the whole test again? 

shall we update the test_start_time again or should we add the question some were for later reference of the student?
@RishabhVerma

**Comment by RishabhVerma** [comment]
@amarkrsinha1997 We can store the generated questions somewhere which will be stored only once. From the next time if this API endpoint is pinged the same set of questions will be returned.

**Comment by amarkrsinha1997** [comment]
I have used api.model to create the fields object in end_test route to get the questions_attempts as a list in the payload because using reqparse to accept a list of question-attempts would have been a real pain to me so I used the api.model for the (questions_attempts) payload handling and validation. Shall we implement the same in other routes that are handling post or let it be with reqparse? @RishabhVerma

**Comment by RishabhVerma** [comment]
what is `api.model`?

**Comment by amarkrsinha1997** [comment]
It was used to create the response objects using ```fields.String``` and all and in the rest plus it also supports the payload handling by serializing the data according to the object created by the model.

**Comment by amarkrsinha1997** [comment]
Note: Need to create some function to check if the posted question_ids exist or not with the option_ids provided and if it doesn't exist then send the list of ids which doesn't exist with an error message and the question_attempts shall not be submitted

Make a string of question set that has been selected for the student to send him the question and save it in an enrollment table using a new field questions_set

**Comment by RishabhVerma** [comment]
> Note: Need to create some function to check if the posted question_ids exist or not with the option_ids provided and if it doesn't exist then send the list of ids which doesn't exist with an error message and the question_attempts shall not be submitted

> Make a string of question set that has been selected for the student to send him the question and save it in an enrollment table using a new field questions_set

Sounds perfect.
